crates/chain/src/keychain/txout_index.rs

```diff
@@ -803,50 +665,67 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-        self.reveal_to_target_with_id(descriptor_id, target_index)
+    ) -> Option<(Vec<Indexed<ScriptBuf>>, ChangeSet<K>)> {
+        let mut changeset = ChangeSet::default();
+        let mut spks: Vec<Indexed<ScriptBuf>> = vec![];
+        while let Some((i, new)) = self.next_index(keychain) {
+            if !new || i > target_index {
+                break;
+            }
+            match self.reveal_next_spk(keychain) {
+                Some(((i, spk), change)) => {
+                    spks.push((i, spk));
+                    changeset.append(change);
+                }
+                None => break,
+            }
+        }
+
+        Some((spks, changeset))
     }
 
     /// Attempts to reveal the next script pubkey for `keychain`.
     ///
     /// Returns the derivation index of the revealed script pubkey, the revealed script pubkey and a
-    /// [`super::ChangeSet`] which represents changes in the last revealed index (if any).
+    /// [`ChangeSet`] which represents changes in the last revealed index (if any).
     /// Returns None if the provided keychain doesn't exist.
     ///
     /// When a new script cannot be revealed, we return the last revealed script and an empty
-    /// [`super::ChangeSet`]. There are two scenarios when a new script pubkey cannot be derived:
+    /// [`ChangeSet`]. There are two scenarios when a new script pubkey cannot be derived:
     ///
     ///  1. The descriptor has no wildcard and already has one script revealed.
     ///  2. The descriptor has already revealed scripts up to the numeric bound.
     ///  3. There is no descriptor associated with the given keychain.
-    pub fn reveal_next_spk(
-        &mut self,
-        keychain: &K,
-    ) -> Option<((u32, &Script), super::ChangeSet<K>)> {
-        let descriptor_id = self.keychains_to_descriptors.get(keychain)?.0;
-        let (next_index, _) = self.next_index(keychain).expect("We know keychain exists");
-        let changeset = self
-            .reveal_to_target(keychain, next_index)
-            .expect("We know keychain exists")
-            .1;
+    pub fn reveal_next_spk(&mut self, keychain: &K) -> Option<(Indexed<ScriptBuf>, ChangeSet<K>)> {
+        let (next_index, new) = self.next_index(keychain)?;
+        let mut changeset = ChangeSet::default();
+
+        if new {
+            let did = self.keychains_to_descriptor_ids.get(keychain)?;
+            let descriptor = self.descriptor_ids_to_descriptors.get(did)?;
+            let spk = descriptor
+                .at_derivation_index(next_index)
+                .expect("already checked index is not too high")
+                .script_pubkey();
+            let _ = self.inner.insert_spk((keychain.clone(), next_index), spk);
+            self.last_revealed.insert(*did, next_index);
+            changeset.last_revealed.insert(*did, next_index);
+            self.replenish_lookahead(keychain, self.lookahead);
```

Comment:
I'm thinking as long as we update `self.last_revealed` and then call `replenish_lookahead`, we shouldn't have to insert the spk here.

Reply:
Ah nice catch. Hmm I would then change the name of the `lookahead` functions so that they are not just for lookaheads.

see 1b6fe8cb0ae0104dac81730e34972804c04114b5